File: blokus-engine/tests/test_pieces.py
"""Tests for the pieces module."""

import pytest
from blokus.pieces import (
    Piece, PieceType, PIECES, PIECE_SHAPES,
    get_piece, get_all_pieces, num_orientations,
    _normalize_coords, _rotate_90, _flip_horizontal
)


class TestPieceShapes:
    """Test piece definitions."""
    
    def test_all_21_pieces_defined(self):
        """Verify all 21 pieces are defined."""
        assert len(PieceType) == 21
        assert len(PIECE_SHAPES) == 21
        assert len(PIECES) == 21
    
    def test_piece_sizes(self):
        """Verify each piece has correct number of squares."""
        expected_sizes = {
            PieceType.I1: 1,
            PieceType.I2: 2,
            PieceType.I3: 3, PieceType.L3: 3,
            PieceType.I4: 4, PieceType.L4: 4, PieceType.T4: 4, PieceType.O4: 4, PieceType.S4: 4,
            PieceType.F: 5, PieceType.I5: 5, PieceType.L5: 5, PieceType.N: 5,
            PieceType.P: 5, PieceType.T5: 5, PieceType.U: 5, PieceType.V: 5,
            PieceType.W: 5, PieceType.X: 5, PieceType.Y: 5, PieceType.Z: 5,
        }
        
        for piece_type, expected_size in expected_sizes.items():
            piece = get_piece(piece_type)
            assert piece.size == expected_size, f"{piece_type.value} should have {expected_size} squares"
    
    def test_total_squares(self):
        """Verify total squares per player is 89."""
        total = sum(get_piece(pt).size for pt in PieceType)
        assert total == 89


class TestPieceOrientations:
    """Test piece rotations and orientations."""
    
    def test_monomino_has_one_orientation(self):
        """Monomino (single square) has only 1 orientation."""
        assert num_orientations(PieceType.I1) == 1
    
    def test_domino_has_two_orientations(self):
        """Domino has 2 orientations (horizontal/vertical)."""
        assert num_orientations(PieceType.I2) == 2
    
    def test_square_has_one_orientation(self):
        """O4 (square) has only 1 orientation."""
        assert num_orientations(PieceType.O4) == 1
    
    def test_x_has_one_orientation(self):
        """X pentomino (plus sign) has only 1 orientation."""
        assert num_orientations(PieceType.X) == 1
    
    def test_asymmetric_pieces_have_eight_orientations(self):
        """Asymmetric pieces like F, N, Y have 8 orientations. Z has 4 (rotational symmetry)."""
        for pt in [PieceType.F, PieceType.N, PieceType.Y]:
            assert num_orientations(pt) == 8, f"{pt.value} should have 8 orientations"
        # Z has rotational symmetry, so only 4 unique orientations
        assert num_orientations(PieceType.Z) == 4, "Z should have 4 orientations"
    
    def test_all_orientations_unique(self):
        """All orientations of a piece should be unique."""
        for piece_type in PieceType:
            orientations = PIECES[piece_type]
            coords_set = {piece.coords for piece in orientations}
            assert len(coords_set) == len(orientations), f"{piece_type.value} has duplicate orientations"


class TestPieceTransformations:
    """Test rotation and flip operations."""
    
    def test_normalize_coords(self):
        """Normalization should shift to origin."""
        coords = [(2, 3), (2, 4), (3, 3)]
        normalized = _normalize_coords(coords)
        assert normalized == frozenset([(0, 0), (0, 1), (1, 0)])
    
    def test_rotate_90_single(self):
        """Test 90 degree rotation."""
        # Horizontal line -> vertical line
        coords = frozenset([(0, 0), (0, 1), (0, 2)])
        rotated = _rotate_90(coords)
        expected = frozenset([(0, 0), (1, 0), (2, 0)])
        assert rotated == expected
    
    def test_four_rotations_returns_to_original(self):
        """Four 90° rotations should return to original."""
        original = frozenset([(0, 0), (0, 1), (1, 0)])
        current = original
        for _ in range(4):
            current = _rotate_90(current)
        assert current == original
    
    def test_flip_horizontal(self):
        """Test horizontal flip."""
        # L shape
        coords = frozenset([(0, 0), (0, 1), (1, 0)])
        flipped = _flip_horizontal(coords)
        expected = frozenset([(0, 0), (0, 1), (1, 1)])
        assert flipped == expected


class TestPieceCorners:
    """Test corner detection for placement rules."""
    
    def test_monomino_corners(self):
        """Monomino has 4 corner positions."""
        piece = get_piece(PieceType.I1)
        corners = piece.get_corners()
        assert len(corners) == 4
        expected = {(-1, -1), (-1, 1), (1, -1), (1, 1)}
        assert corners == expected
    
    def test_domino_corners(self):
        """Domino has 4 corner positions (at ends)."""
        piece = get_piece(PieceType.I2)
        corners = piece.get_corners()
        # Should have corners only at the ends, not in the middle
        assert len(corners) == 4
    
    def test_corners_not_edge_adjacent(self):
        """Corners should not be edge-adjacent to any piece square."""
        for piece_type in PieceType:
            piece = get_piece(piece_type)
            corners = piece.get_corners()
            for corner in corners:
                for r, c in piece.coords:
                    # Manhattan distance should be > 1 (not edge adjacent)
                    dist = abs(corner[0] - r) + abs(corner[1] - c)
                    assert dist > 1, f"Corner {corner} is edge-adjacent to ({r},{c}) in {piece_type.value}"


class TestPieceEdges:
    """Test edge detection."""
    
    def test_monomino_edges(self):
        """Monomino has 4 edge positions."""
        piece = get_piece(PieceType.I1)
        edges = piece.get_edges()
        assert len(edges) == 4
        expected = {(-1, 0), (1, 0), (0, -1), (0, 1)}
        assert edges == expected


class TestPieceMatrix:
    """Test matrix conversion."""
    
    def test_monomino_matrix(self):
        """Monomino should be 1x1 matrix."""
        piece = get_piece(PieceType.I1)
        matrix = piece.to_matrix()
        assert matrix.shape == (1, 1)
        assert matrix[0, 0] == 1
    
    def test_domino_matrix(self):
        """Domino should be 1x2 or 2x1 matrix depending on orientation."""
        piece = get_piece(PieceType.I2, orientation=0)
        matrix = piece.to_matrix()
        assert matrix.sum() == 2
        assert 1 in matrix.shape and 2 in matrix.shape

File: blokus-web/tests/test-pieces.js
/**
 * Tests for pieces.js module
 */

import { 
    describe, test, assert, assertEqual, assertNotEqual 
} from './test-framework.js';

import { 
    PieceType, Piece, PIECES, 
    getPiece, getOrientationCount, getAllPieceTypes,
    normalizeCoords, rotate90, flipHorizontal
} from '../js/pieces.js';

// ============================================================================
// PIECE DEFINITIONS TESTS
// ============================================================================

describe('PieceType', () => {
    test('all 21 pieces are defined', () => {
        const types = getAllPieceTypes();
        assertEqual(types.length, 21, 'Should have 21 piece types');
    });
    
    test('PieceType is frozen', () => {
        assert(Object.isFrozen(PieceType), 'PieceType should be frozen');
    });
});

describe('Piece sizes', () => {
    test('monomino has 1 square', () => {
        assertEqual(getPiece(PieceType.I1).size, 1, 'I1 should have 1 square');
    });
    
    test('domino has 2 squares', () => {
        assertEqual(getPiece(PieceType.I2).size, 2, 'I2 should have 2 squares');
    });
    
    test('triminoes have 3 squares', () => {
        assertEqual(getPiece(PieceType.I3).size, 3, 'I3 should have 3 squares');
        assertEqual(getPiece(PieceType.L3).size, 3, 'L3 should have 3 squares');
    });
    
    test('tetrominoes have 4 squares', () => {
        for (const type of [PieceType.I4, PieceType.L4, PieceType.T4, PieceType.O4, PieceType.S4]) {
            assertEqual(getPiece(type).size, 4, `${type} should have 4 squares`);
        }
    });
    
    test('pentominoes have 5 squares', () => {
        const pentominoes = [
            PieceType.F, PieceType.I5, PieceType.L5, PieceType.N, PieceType.P,
            PieceType.T5, PieceType.U, PieceType.V, PieceType.W, PieceType.X, 
            PieceType.Y, PieceType.Z
        ];
        for (const type of pentominoes) {
            assertEqual(getPiece(type).size, 5, `${type} should have 5 squares`);
        }
    });
    
    test('total squares per player is 89', () => {
        let total = 0;
        for (const type of getAllPieceTypes()) {
            total += getPiece(type).size;
        }
        assertEqual(total, 89, 'Total squares should be 89');
    });
});

// ============================================================================
// ORIENTATION TESTS
// ============================================================================

describe('Piece orientations', () => {
    test('monomino has 1 orientation', () => {
        assertEqual(getOrientationCount(PieceType.I1), 1, 'I1 should have 1 orientation');
    });
    
    test('domino has 2 orientations', () => {
        assertEqual(getOrientationCount(PieceType.I2), 2, 'I2 should have 2 orientations');
    });
    
    test('square (O4) has 1 orientation', () => {
        assertEqual(getOrientationCount(PieceType.O4), 1, 'O4 should have 1 orientation');
    });
    
    test('X pentomino has 1 orientation', () => {
        assertEqual(getOrientationCount(PieceType.X), 1, 'X should have 1 orientation');
    });
    
    test('asymmetric pieces have 8 orientations', () => {
        for (const type of [PieceType.F, PieceType.N, PieceType.Y]) {
            assertEqual(getOrientationCount(type), 8, `${type} should have 8 orientations`);
        }
    });
    
    test('all orientations are unique', () => {
        for (const type of getAllPieceTypes()) {
            const orientations = PIECES[type];
            const seen = new Set();
            for (const piece of orientations) {
                const key = JSON.stringify([...piece.coords].sort());
                assert(!seen.has(key), `${type} should have unique orientations`);
                seen.add(key);
            }
        }
    });
});

// ============================================================================
// TRANSFORMATION TESTS
// ============================================================================

describe('Coordinate transformations', () => {
    test('normalizeCoords shifts to origin', () => {
        const coords = [[2, 3], [2, 4], [3, 3]];
        const normalized = normalizeCoords(coords);
        assertEqual(normalized, [[0, 0], [0, 1], [1, 0]], 'Should normalize to origin');
    });
    
    test('rotate90 rotates clockwise', () => {
        // Horizontal line -> vertical line
        const coords = [[0, 0], [0, 1], [0, 2]];
        const rotated = rotate90(coords);
        assertEqual(rotated.length, 3, 'Should have same number of cells');
        // Check it's now vertical
        const cols = new Set(rotated.map(c => c[1]));
        assertEqual(cols.size, 1, 'After rotation, all should have same column');
    });
    
    test('four rotations return to original', () => {
        const original = [[0, 0], [0, 1], [1, 0]];
        let current = normalizeCoords(original);
        for (let i = 0; i < 4; i++) {
            current = rotate90(current);
        }
        const originalKey = JSON.stringify(normalizeCoords(original).sort());
        const currentKey = JSON.stringify(current.sort());
        assertEqual(currentKey, originalKey, '4 rotations should return to original');
    });
    
    test('flipHorizontal mirrors piece', () => {
        const coords = [[0, 0], [0, 1], [1, 0]];
        const flipped = flipHorizontal(coords);
        // Should be mirrored
        assertNotEqual(
            JSON.stringify(flipped.sort()), 
            JSON.stringify(coords.sort()),
            'Flipped should be different from original'
        );
    });
});

// ============================================================================
// PIECE METHODS TESTS
// ============================================================================

describe('Piece methods', () => {
    test('getBoundingBox returns correct dimensions', () => {
        const piece = getPiece(PieceType.I5, 0);
        const bbox = piece.getBoundingBox();
        // I5 is either 5x1 or 1x5 depending on orientation
        assert(
            (bbox.height === 5 && bbox.width === 1) || (bbox.height === 1 && bbox.width === 5),
            'I5 should be 5x1 or 1x5'
        );
    });
    
    test('translate moves coordinates by offset', () => {
        const piece = getPiece(PieceType.I1);
        const translated = piece.translate(5, 3);
        assertEqual(translated[0][0], 5, 'Row should be offset');
        assertEqual(translated[0][1], 3, 'Col should be offset');
    });
    
    test('getCorners returns diagonal positions', () => {
        const piece = getPiece(PieceType.I1);
        const corners = piece.getCorners();
        assertEqual(corners.length, 4, 'Monomino should have 4 corners');
    });
    
    test('getEdges returns adjacent positions', () => {
        const piece = getPiece(PieceType.I1);
        const edges = piece.getEdges();
        assertEqual(edges.length, 4, 'Monomino should have 4 edges');
    });
    
    test('corners are not edge-adjacent', () => {
        for (const type of getAllPieceTypes()) {
            const piece = getPiece(type);
            const corners = piece.getCorners();
            for (const [cr, cc] of corners) {
                for (const [pr, pc] of piece.coords) {
                    const dist = Math.abs(cr - pr) + Math.abs(cc - pc);
                    assert(dist > 1, `Corner should not be edge-adjacent in ${type}`);
                }
            }
        }
    });
});

// ============================================================================
// IMMUTABILITY TESTS
// ============================================================================

describe('Immutability', () => {
    test('PIECES is frozen', () => {
        assert(Object.isFrozen(PIECES), 'PIECES should be frozen');
    });
    
    test('Piece.coords is frozen', () => {
        const piece = getPiece(PieceType.I1);
        assert(Object.isFrozen(piece.coords), 'Piece.coords should be frozen');
    });
});

export function runPiecesTests() {
    console.log('\n🧪 Running pieces.js tests...');
}
